catalyst-gateway-crates/c509-certificate/examples/cli/main.rs

```diff
@@ -0,0 +1,350 @@
+        c509_json.issuer,
+        c509_json.subject.clone(),
+    )?;
+
+    // Parse validity dates or use defaults
+    // Now for not_before date
+    let not_before = parse_or_default_date(c509_json.validity_not_before, Utc::now().timestamp())?;
+    // Default as expire date for not_after
+    // Expire date = 9999-12-31T23:59:59+00:00 as mention in the C509 document
+    let not_after = parse_or_default_date(
+        c509_json.validity_not_after,
+        parse_or_default_date(Some("9999-12-31T23:59:59+00:00".to_string()), 0)?,
+    )?;
+
+    let public_key = parse_public_key(&c509_json.subject_public_key)?;
+
+    let key_type = get_key_type(key_type)?;
+
+    // Create TbsCert instance
+    let tbs = TbsCert::new(
+        c509_json.certificate_type.unwrap_or(SELF_SIGNED_INT),
+        serial_number,
+        Name::new(NameValue::RelativeDistinguishedName(issuer)),
+        Time::new(not_before),
+        Time::new(not_after),
+        Name::new(NameValue::RelativeDistinguishedName(c509_json.subject)),
+        c509_json
+            .subject_public_key_algorithm
+            .unwrap_or(SubjectPubKeyAlgorithm::new(key_type.0.clone(), key_type.1)),
+        public_key.to_bytes(),
+        c509_json.extensions.clone(),
+        c509_json
+            .issuer_signature_algorithm
+            .unwrap_or(IssuerSignatureAlgorithm::new(key_type.0, ED25519.1)),
+    );
+
+    let cert = c509_certificate::generate(&tbs, private_key)?;
+
+    // If the output path is provided, write to the file
+    if let Some(output) = output {
+        write_to_output_file(output, &cert)?;
+    };
+
+    println!("Hex: {:?}", hex::encode(&cert));
+    println!("Bytes: {:?}", &cert);
+
+    Ok(())
+}
+
+/// Write a data to a file given an output path.
+fn write_to_output_file(output: PathBuf, data: &[u8]) -> anyhow::Result<()> {
+    let mut file = File::create(output).map_err(|e| anyhow::anyhow!(e))?;
+    file.write_all(data).map_err(|e| anyhow::anyhow!(e))?;
+    Ok(())
+}
+
+/// Determine issuer of the certificate.
+/// If self-signed is true, issuer is the same as subject.
+/// Otherwise, issuer should be present.
+fn determine_issuer(
+    self_signed: bool, issuer: Option<RelativeDistinguishedName>,
+    subject: RelativeDistinguishedName,
+) -> anyhow::Result<RelativeDistinguishedName> {
+    if self_signed {
+        Ok(subject)
+    } else {
+        issuer.ok_or_else(|| anyhow::anyhow!("Issuer should be present if self-signed is false"))
+    }
+}
+
+/// Validate the certificate type.
+fn validate_certificate_type(
+    self_signed: bool, certificate_type: Option<u8>,
+) -> anyhow::Result<()> {
+    if self_signed && certificate_type.unwrap_or(SELF_SIGNED_INT) != SELF_SIGNED_INT {
+        return Err(anyhow::anyhow!(
+            "Certificate type should be 0 if self-signed is true"
+        ));
+    }
+    Ok(())
+}
+
+/// Parse public key from file path or string.
+fn parse_public_key(public_key: &str) -> anyhow::Result<PublicKey> {
```

Comment:
I am not quite sure that this function will that way as you expected.
I would rather explicitly choose at the place or instantiating `PublicKey`  is it generated from file or from string.